storage: log when slow operations terminate

tbg opened the pull request:
Log when a slow proposal, lease acquisition, or quota pool acquisition terminates.
Previously, we'd only log once when a timer had expired.

This is helpful to correlate the end times with other events across the cluster.
For example, if a slow command blocks the command queue, knowing approximately
when it leaves the command queue can help reason about other commands queued
behind it.
Similar, for slow lease proposals it's interesting when exactly they unwedge, and
whether that correlates with any other activity on the range such as snapshots.

Prompted specifically by #28179, where we see that leases for r1 are getting stuck,
but we don't even really know when they get unstuck.

Release note: None

tbg committed:
storage: log actual duration for slow lease proposals

Release note: None

tbg committed:
storage: log actual duration on slow quota pool

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28519)
<!-- Reviewable:end -->

tbg committed:
storage: log actual duration of slow proposal

The call to timeutil.Now() in the hot path could be optimized away,
but the resulting code would be clunky.

Release note: None

a-robinson reviewed (approved):
LGTM I suppose since I understand the motivation, although I'm a little worried about a thundering herd of log messages from tens of thousands of waiting goroutines using up precious cpu/disk resources all at once just as a cluster is starting to recover.

bdarnell reviewed (approved):
> Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28519)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tschottdorf commented:
@a-robinson I thought about adding `log.Every`s for these, but I've found that sometimes I want to find events on a specific rangeID, and it'd be a bummer to have to think about whether the messages might've been suppressed. You may have a point about the congestion, though I'm not convinced that it would be more than a blip compared to the actual termination of the piled up work.

bors r=bdarnell,a-robinson
